### ISSUE TITLE
Disjoint union switch detection is broken for string constants 

### ISSUE THREAD

**Issue by entropitor** [facebook/flow#5083]
Works fine with literal strings: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVxgBcCeAHAUzAHk8sBLOAOzAF4wBvMXQgLjAHIBBDsAXzAAfRs3wF2HAEK8Bwpi3GcAwjIDcqBSTKUa9br2FSDyjuigBXKgGMK1MAEMAzo4IAnLADkCANzcAKcgBbPDhncgAjGEUCYNwASnZ7KhxGVD4zSxsdMDxXOCsCZ1JbKj84dmKdBLAqc0Dwt1SwMEcEciwrAAswMrim5rArJyJ9dgZUAYHXAixzVxoARnVJ9Mmhl05pMYnJsGnZ+bAAJmWB1YH1kZVt3b2ZuZoTnebz5oATAih7cxgsG93MA4YDAwHBtNRHIM4L5pm8ADSDJJUOBYMCwRDMOBgTpuIgAWjAADF4Eg3nBCjUUQ4AObTAgAQmeU3uhycLncXhhvVOLx26T4QA

Doesn't work with string constants: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVxgBcCeAHAUzAHk8sBLOAOzAF4wBvMXQgLjAHIBBDsAXzAAfRs3wF2HAEK8Bwpi3GcAwjIDcqAgA88cAE5YwAY2oBnA1wk86nHhu16DxqmbCSJ061I52d+o6YMlCRVPFQxgUUISMkoaem5eYS8hZW8FaIpqawU4KDAuFJy8yUKxXLAldCgAVypDTJoAQxMTAn0AOQIANzaACnIAWx0W8gAjGEUCIdwASnZGqhxGVD4q2vrYsDxdOEMCFtIG3rh2Q9i5sCpqgdG25bAwEwRyLEMACzBjmfuHo2aiCw-X5gXQELDVXQ0ACM6mBq2Bhn+rnYDFQwIeoPBkLAACZYb94b9Ea0Kii0ejMRCaHjyQ9CQ8ACYEKCNaowLBk9FgTBgRowGBgOAxUz+HqghkAGj+VCocAMsEQzDgYDebSIAFowAAxeBIBlwfaXOW8gDmoIIAEJab9Kdjmq0Ot0+nAZvi6eTVnwgA

**Comment by chrisgervang** [comment]
It doesn't catch typos in the switch statement for string literals in your example when using your `type Option = { type: 'A' }`, either.

Compare [example doesn't catch typo](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDyOGAlnAHZgC8YA3mNvgFxgDkAgm2AL5gA+9Rrjys2AIW59BDJqPYBhKQG5UwYHKIlyVWp26CJBxW3RQArhQDGZSmACGAZ0d4AThgByeAG5uAFKQAtjhwzqQARjDyeMHYAJSs9hRY9Kg8ZpY2OmA4rnBWeM7EthR+cKzFOglgFOaB4W6pYGCOCKQYVgAWYGUAdHJxTc1gVk4EnAD6bnmubKx0qMPDrngY5q5UAIyqS+lLoy7skvOLS2AraxtgAEw7w3vDB+NKJ2fnq+tUt6fND80AJngoPZzDAMK8zuoHDAYGA4NpKI4RnBfCt-gAaEZJChwDBgWCIRhwMCdNwEAC0YAAYvAkP84IUariHABzFZ4ACEP2WHyuThc7i8qLKcTuv1O6R4QA) against [exmaple does catch typo](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVxgBcCeAHAUzAHk8sBLOAOzAF4wBvMXQgLjAHIBBDsAXzAAfRs3wF2HAEK8Bwpi3GcAwjIDcqBSTKUa9br2FSDyjuigBXKgGMK1MAEMAzo4IAnLADkCANzcAKcgBbPDhncgAjGEUCYNwASnZ7KhxGVD4zSxsdMDxXOCsCZ1JbKj84dmKdBLAqc0Dwt1SwMEcEciwrAAswMrim5rArJyJuAH03PNcOdgZUAYHXAixzVxoARnV59Pmhl05pGbn5sEXl1bAAJk2B7YHdkZVD45OllZoro+bb5oATAih7OYYFgnsdMA4YDAwHBtNRHIM4L5Fj8ADSDJJUOBYMCwRDMOBgTpuIgAWjAADF4EgfnBCjUsQ4AOaLAgAQk+C1e5ycLncXiRvWuXyO6T4QA).

Are there plans to support this type checking for discriminated unions, [similar to typescript](https://basarat.gitbooks.io/typescript/docs/types/discriminated-unions.html)?

**Comment by mkscrg** [comment]
Here's a pattern I've been using to catch typos on union switches (in Redux reducers 😝) ([TryFlow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDyOGAlnAHZgC8YA3mNvgFxgDkAgm2AL5gA+9Rrjys2AIW48A3JhFES5CgBV5tACQAFAE5x827KvwAeYmUoAaMACImeawD5ZqKAFcKAY3NUcuj3gBnALMlAAo4VhDKAEpWClcAWwAjPG16VDAwD0oAjDA4I1EFb0KafIA6O1lMgIRSDA8ACzBwwuj0zMyPAEMAgk4AfVTdbTZWOgzOzO08DFdtKgBGas6eSa7e-slx9c6ZuYWwACYVzLWpgBM8KG7XGAwdqczWkVY8BJIsaNO92fmqAAMP3OvFQPCAA)):

```js
type Option = { type: 'A' } | { type: 'B' };
type OptionType = $PropertyType<Option, "type">;

function processOption(o: Option): number {
  const oType: OptionType = o.type;
  switch (oType) {
    case 'A_error': {
      return 1;
    }
    case 'B': {
      return 2;
    }
    default: {
      (oType: empty);
      return 0;
    }
  }
}
```

This at least generates a compiler error at `'A_error'`, though it's not Flow's most readable error 🙄

**Comment by entropitor** [comment]
@mkscrg Have you tried your example with string constants? Because that's what this issue is about. It works without string constants, but not with them. 

Off course full support for switches and disjoint union would be nice but there are already other issues open for that.

**Comment by mkscrg** [comment]
Aha! My bad @entropitor, I misunderstood your description. Indeed this is a separate issue.

**Comment by daryn-k** [comment]
I have a same problem. What is the solution?

**Comment by nrabinowitz** [comment]
Same issue - here's an [even clearer repro case](https://flow.org/try/#0C4TwDgpgBAygtgQwDZIKoDsCWB7dUC8UA5AGbbZFQA+xARggE5EDcAUKJFAEKYDmGOPIXjI0WXNToIAXpRroArijasAxrgDOwKADEA8noLEyFNuvRbuAQQBKRovSYqSC9KuCCo5gG4QGwHQRMJAAKDggALm4+AVwASiiRFFi8eSUkKABvVigoDQB3TGBVAAsoMPAIOKyc3K8EDWh9PQja3NUG6C5bVrq6hghgBQY8cLY+gBMIEgQlYF6+qAGhkahFZVqAX1Zt1hc3DwkfP2AYBVVVCA0NCsgonn5xdATYRGSnyXWM7NyCotLyuFqj86h1GsZyEQFu1OlImND+oNhqNKuM6lMZnMEbllsi1uk0dtNkA).

**Comment by nrabinowitz** [comment]
Oddly, this [works with constants if only the last case is a string literal.](https://flow.org/try/#0C4TwDgpgBAygtgQwDZIKoDsCWB7dUC8UA5AGbbZFQA+xARggE5EDcAUKJFAEKYDmGOPIXjI0WXNToIAXpRroArijasAxrgDOwKADEA8noLEyFNuvRbuAQQBKRovSYqSC9KuCCo5gG4QGwHQRMJAAKDggALm4+AVwASiiRFFi8eSUkKABvVigoDQB3TGBVAAsoMPAIOKyc3K8EDWh9PQja3NUG6AdGIla6uoYIYAUGPHC2foATCBIEJWA+-qhB4dGoRWVagF9WHaA)

**Comment by vladar** [comment]
Similar (but slightly different) issue with disjoint unions and constants: [repro](https://flow.org/try/#0PTAEAEDMBsHsHcBQBjWA7AzgF1AMQPL6gC8oA5AfmQNwrragBCAggEonkus12Y4sAtDmUE9EWAJ4AHAKZ5YsZsiwBLdBwDeiUKEmyAXOUpkANNtAATAIZYrh7ACcVaAOZmAvonHS5jKw6VVdVItHT0ZQxE2U3NrW3ssJ1cPL3CmKwAvQLU0TXNwyNEzHTi7UEdnN0RPb1lQbODzAB95RWUc5vSA9vREEB0WvyyetERaRBkADylYBxwLGUgrAFdoHEhltBHQBxkLZeQZBwAKcx1sGwjyxMrhGJ0dKxHDBtGASgSkl1BQ8vgVLDIAAWoGOTyCaAAdOE3j8zqBkFYMHJKPp4YjkUw2GiHg9dlhlg5cuCcpDSuNcRjfMwBDjcTsZASiaASegyTYrBSHgslqssHTcWDnqAZABbKSSN60ekMpm5C5YGTS0CeTxAA) (expecting "all branches are incompatible" error there). The same case with [hardcoded strings](https://flow.org/try/#0PTAEAEDMBsHsHcBQBjWA7AzgF1AMQPL6gC8oA5AfmQNwrragBCAggEonkus12Y4sAtDmUE9EWAJ4AHAKZ5YsZsiwBLdBwDeiUKEmyAXOUpkANNtAATAIZYrh7ACcVaAOZmAvonHS5jKw6VVdVItHT0ZQxE2U3NrW3ssJ1cPL3CmKwAvQLU0TXNwyNEzHTi7UEdnN0RPb1lQbODzAB95RWUc5vSA9vREEB0WvyyetERaRBkADylYBxwLGUgrAFdoHEhltBHQBxkLZeQZBwAKcx1sGwjyxMrhGJ0dKxHDBtGASgSkl1BQ8vgVLDIAAWoGOTyCaAAdOE3j8zqBkFYMHJKPp4YjkUw2GiHg9dlhlg5cuCcpDSuNcRi5FEBGQcbidjICUTQCT0GSbFYKQ8FktVlh6biwc9QDIALZSSRvWgMxnM3IXLAyGWgTyeIA) - works as expected (produces an error)

**Comment by gkz** [comment]
Fixed by 658e536a4d6d4ffc38b5e76b22eac9c087a958a9
